### DISCUSSION TITLE
SSR support for non Javascript projects

### DISCUSSION THREAD

**Discussion by jefer94** [lit/lit#4475] [General]
I built an API to bring SSR to frameworks like Django, I think [renderLitElement](https://github.com/jefer94/lit-ssr-api/blob/73bdaf74a7903ca12be3813ffcc92eda2880675d/utils/renderLitElement.ts#L60) should be used over a Lambda Function or you should deploy the [service](https://github.com/jefer94/lit-ssr-api) and use it in different projects.

I think that the reason #1 to use this is to add support to Web Components from your unsupported backed

**Comment by jefer94** [comment]
https://github.com/lit/lit/discussions/2768

**Comment by augustjk** [comment]
Aha that's what you meant by an API. That's an interesting solution! The main concern I would have is the performance cost of using a separate service for SSR-ing parts of your page.

Spawning a local node process to do the SSR could be faster than reaching out to another service, but perhaps that would depend on your hardware/infrastructure.

For a Go backend, this might be another solution https://github.com/lit/lit/discussions/4193

**Comment by jefer94** [comment]
Mmmm interesting, I will make some benchmarks, I was making an API thinking in the deduplication but I guess that becomes it a NPM dep should cover this purpose

**Comment by jefer94** [comment]
I'm building an NPM package to be reused by a server or an RPC implementation, it should be used until a better implementation allows SSR without using a Node process as a proxy
https://github.com/jefer94/lit-ssr-utils/tree/main

I got to implement a Unix socket, which is significantly faster than an API Rest, it serializes the data with Protobuf
- https://github.com/jefer94/lit-ssr-utils/blob/main/examples/client/unix-domain-socket/index.ts
- https://github.com/jefer94/lit-ssr-utils/blob/main/examples/server/unix-domain-socket/index.ts

It exports 3 apis
- Base for API Rest https://github.com/jefer94/lit-ssr-utils/blob/main/src/api.ts
- Base for a command https://github.com/jefer94/lit-ssr-utils/blob/main/src/render.ts (it will be remove, this idea has no sense)
- Base for RPC https://github.com/jefer94/lit-ssr-utils/blob/main/src/rpc.ts

Important:
- This process will grow its size in memory, and it is necessary to design a strategy to release the memory, maybe implement something like a supervisor that kills and respawns the process after N requests
- It is necessary to design the API of the library, I'm implementing a client/server architecture 
- Maybe the package should have a better name
- Does it support Windows devs? maybe should be necessary an API Rest to cover the development environment
- The process uses 108.5MB of RAM, should it use less memory?
- I cannot finish the implementation using `Bun.serve`, maybe this should improve the performance
- Some examples does not works yet

I have time to make changes next week

**Comment by jefer94** [comment]
Good news, I published [lit-ssr-utils@v.0.2.3](https://www.npmjs.com/package/lit-ssr-utils), coming soon I will publish `lit_ssr` for Python.

I am waiting for this https://github.com/oven-sh/bun/issues/8044 to enable Unix sockets.

**Reply by MentalGear** [reply]
Interesting. Could you also make this work for static generation for a build tool like **vite**?

**Reply by jefer94** [reply]
The solution that I thought was to save the static content in Redis, but it technically should generate the static files, but my question would be why don't use Astro? It is even an agnostic tool that you should use with React, Solid, Angular and Lit, so, if you tell me what would be the differences I could work in it.

I was working in this library, but I had a lot of important things to do in my work.
> https://github.com/jefer94/lit-ssr.py

The problem that I tried to solve was, I have many Javascript libraries what can pre render the views, but I have the models within Django, so I need or... send the required data to Javacript or send the pre rendered HTML to Javascript using Django template to fill the content.

I have an issue that I have my monolith separate into services, use Astro in this case is complicated, but if your API does not have this issue, you should send the Astro dist folder to your public folder in your favorite framework.

You should see that I have a file to dynamically load the HTTP client that you chose, and then pass the argument thought the network, but it is not completed yet.
1. https://github.com/jefer94/lit-ssr.py/blob/main/src/lit_ssr/rest/__init__.py
2. https://github.com/jefer94/lit-ssr.py/blob/main/src/lit_ssr/rest/aiohttp.py

**Reply by jefer94** [reply]
Another stuff that should be fixed, my lib sends the whole dependencies thought the network, but the last version was designed to work aside of your main backend to reduce the latency (a non Kubernetes environment) it technically should send just the path and Javascript should load the file directly.
